pkg/kubelet/kubelet_server_journal.go

```diff
@@ -0,0 +1,431 @@
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(30*time.Second))
+	defer cancel()
+	boot := int(0)
+	if n.Boot != nil {
+		boot = *n.Boot
+	}
+	n.copyForBoot(ctx, w, boot)
+}
+
+// copyForBoot invokes the OS specific logging command with the  provided args
+func (n *nodeLogQuery) copyForBoot(ctx context.Context, w io.Writer, previousBoot int) {
+	if ctx.Err() != nil {
+		return
+	}
+	nativeLoggers, fileLoggers := n.splitNativeVsFileLoggers(ctx)
+	if len(nativeLoggers) > 0 {
+		n.copyServiceLogs(ctx, w, nativeLoggers, previousBoot)
+	}
+
+	if len(fileLoggers) > 0 && n.options != (options{}) {
+		fmt.Fprintf(w, "\noptions present and query resolved to log files for %v\ntry without specifying options\n",
+			fileLoggers)
+		return
+	}
+
+	if len(fileLoggers) > 0 {
+		copyFileLogs(ctx, w, fileLoggers)
+	}
+}
+
+// splitNativeVsFileLoggers checks if each service logs to native OS logs or to a file and returns a list of services
+// that log natively vs maybe to a file
+func (n *nodeLogQuery) splitNativeVsFileLoggers(ctx context.Context) ([]string, []string) {
+	var nativeLoggers []string
+	var fileLoggers []string
+
+	for _, service := range n.Services {
+		// Check the journalctl output to figure if the service is using journald or not. This is not needed in the
+		// Get-WinEvent case as the command returns an error if a service is not logging to the Application provider.
+		if checkForNativeLogger(ctx, service) {
+			nativeLoggers = append(nativeLoggers, service)
+		} else {
+			fileLoggers = append(fileLoggers, service)
+		}
+	}
+	return nativeLoggers, fileLoggers
+}
+
+// copyServiceLogs invokes journalctl or Get-WinEvent with the provided args. Note that
+// services are explicitly passed here to account for the heuristics.
+func (n *nodeLogQuery) copyServiceLogs(ctx context.Context, w io.Writer, services []string, previousBoot int) {
+	cmdStr, args, err := getLoggingCmd(n, services)
+	if err != nil {
+		fmt.Fprintf(w, "\nfailed to get logging cmd: %v\n", err)
+		return
+	}
+	cmd := exec.CommandContext(ctx, cmdStr, args...)
+	cmd.Stdout = w
+	cmd.Stderr = w
+
+	if err := cmd.Run(); err != nil {
+		if _, ok := err.(*exec.ExitError); ok {
+			return
+		}
+		if previousBoot == 0 {
+			fmt.Fprintf(w, "\nerror: journal output not available\n")
+		}
+	}
+}
+
+// copyFileLogs loops over all the services and attempts to collect the file logs of each service
+func copyFileLogs(ctx context.Context, w io.Writer, services []string) {
+	if ctx.Err() != nil {
+		fmt.Fprintf(w, "\ncontext error: %v\n", ctx.Err())
+		return
+	}
+
+	for _, service := range services {
+		heuristicsCopyFileLogs(ctx, w, service)
+	}
+}
+
+// heuristicsCopyFileLogs attempts to collect logs from either
+// /var/log/service
+// /var/log/service.log or
+// /var/log/service/service.log or
+// in that order stopping on first success.
+func heuristicsCopyFileLogs(ctx context.Context, w io.Writer, service string) {
+	logFiles := [7]string{
```

Comment:
Why is this 7?

Reply:
Good catch. I missed updating this when reducing the heuristics. Fixed.